File: ios/tools/gen-icon.swift
// Generates Loopkeeper's 1024×1024 app icon (opaque, no alpha) — a white "loop" ring with a
// gap being closed by a checkmark, on the brand gradient. Run:
//   swift ios/tools/gen-icon.swift ios/Assets.xcassets/AppIcon.appiconset/icon-1024.png
import AppKit

let side = 1024
let rep = NSBitmapImageRep(
    bitmapDataPlanes: nil, pixelsWide: side, pixelsHigh: side,
    bitsPerSample: 8, samplesPerPixel: 3, hasAlpha: false, isPlanar: false,
    colorSpaceName: .deviceRGB, bytesPerRow: 0, bitsPerPixel: 0
)!

let ctx = NSGraphicsContext(bitmapImageRep: rep)!
NSGraphicsContext.saveGraphicsState()
NSGraphicsContext.current = ctx
let cg = ctx.cgContext

// Brand gradient background.
let space = CGColorSpaceCreateDeviceRGB()
let grad = CGGradient(
    colorsSpace: space,
    colors: [
        NSColor(srgbRed: 0.30, green: 0.36, blue: 0.92, alpha: 1).cgColor,
        NSColor(srgbRed: 0.12, green: 0.50, blue: 0.98, alpha: 1).cgColor,
    ] as CFArray,
    locations: [0, 1]
)!
cg.drawLinearGradient(grad, start: CGPoint(x: 0, y: side), end: CGPoint(x: side, y: 0), options: [])

cg.setStrokeColor(NSColor.white.cgColor)
cg.setLineCap(.round)

// The "loop": a ring with a gap (an open loop).
cg.setLineWidth(74)
cg.addArc(center: CGPoint(x: 512, y: 512), radius: 300, startAngle: .pi * 0.32, endAngle: .pi * 2.18, clockwise: false)
cg.strokePath()

// The "kept": a checkmark closing it.
cg.setLineWidth(74)
cg.move(to: CGPoint(x: 392, y: 520))
cg.addLine(to: CGPoint(x: 476, y: 426))
cg.addLine(to: CGPoint(x: 648, y: 624))
cg.strokePath()

NSGraphicsContext.restoreGraphicsState()

let out = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "icon-1024.png"
let png = rep.representation(using: .png, properties: [:])!
try! png.write(to: URL(fileURLWithPath: out))
print("wrote \(out)")

File: ios/tools/gen-icon.py
#!/usr/bin/env python3
"""Generate Loopkeeper's 1024x1024 app icon (opaque RGB PNG, no deps).

A white "loop" ring closed by a checkmark, on the brand gradient.
Usage: python3 ios/tools/gen-icon.py ios/Assets.xcassets/AppIcon.appiconset/icon-1024.png
"""
import sys
import zlib
import struct

SIDE = 1024
CX = CY = 512.0
R = 300.0
RING_HALF = 38.0
CHECK_HALF = 40.0
# Checkmark in image coords (y down): down-right short stroke, then up-right long stroke.
CHECK = [((372, 528), (470, 628)), ((470, 628), (664, 404))]
C0 = (0.30, 0.36, 0.92)  # top-left
C1 = (0.12, 0.50, 0.98)  # bottom-right


def seg_dist(px, py, a, b):
    ax, ay = a
    bx, by = b
    dx, dy = bx - ax, by - ay
    l2 = dx * dx + dy * dy
    t = 0.0 if l2 == 0 else max(0.0, min(1.0, ((px - ax) * dx + (py - ay) * dy) / l2))
    qx, qy = ax + t * dx, ay + t * dy
    return ((px - qx) ** 2 + (py - qy) ** 2) ** 0.5


def png(path, w, h, rgb):
    def chunk(typ, data):
        return struct.pack(">I", len(data)) + typ + data + struct.pack(">I", zlib.crc32(typ + data) & 0xFFFFFFFF)
    raw = bytearray()
    for y in range(h):
        raw.append(0)
        raw.extend(rgb[y * w * 3:(y + 1) * w * 3])
    with open(path, "wb") as f:
        f.write(b"\x89PNG\r\n\x1a\n")
        f.write(chunk(b"IHDR", struct.pack(">IIBBBBB", w, h, 8, 2, 0, 0, 0)))
        f.write(chunk(b"IDAT", zlib.compress(bytes(raw), 9)))
        f.write(chunk(b"IEND", b""))


def main():
    out = sys.argv[1] if len(sys.argv) > 1 else "icon-1024.png"
    buf = bytearray(SIDE * SIDE * 3)
    i = 0
    for y in range(SIDE):
        for x in range(SIDE):
            dist = ((x - CX) ** 2 + (y - CY) ** 2) ** 0.5
            white = abs(dist - R) <= RING_HALF
            if not white:
                for a, b in CHECK:
                    if seg_dist(x, y, a, b) <= CHECK_HALF:
                        white = True
                        break
            if white:
                r = g = bl = 255
            else:
                t = (x + y) / (2.0 * SIDE)
                r = int((C0[0] + (C1[0] - C0[0]) * t) * 255)
                g = int((C0[1] + (C1[1] - C0[1]) * t) * 255)
                bl = int((C0[2] + (C1[2] - C0[2]) * t) * 255)
            buf[i] = r; buf[i + 1] = g; buf[i + 2] = bl
            i += 3
    png(out, SIDE, SIDE, buf)
    print("wrote", out)


if __name__ == "__main__":
    main()
